overview/cli.md

```diff
@@ -0,0 +1,369 @@
+
+We can read all about `OptionParser` in [the official API documentation](https://crystal-lang.org/api/latest/OptionParser.html). And from there we are one click away from the source code ... the actual proof that it is not magic!
+
+Now, let's run our application. We have two ways [using the compiler](https://crystal-lang.org/reference/using_the_compiler):
+1. [Build the application](https://crystal-lang.org/reference/using_the_compiler/#crystal-build) and then run it.
+2. Compile and [run the application](https://crystal-lang.org/reference/using_the_compiler/#crystal-run), all in one command.
+
+We are going to use the second way:
+
+```shell-session
+$ crystal ./all_my_cli.cr -- -h
+
+Welcome to The Beatles App!
+    -v, --version                    Show version
+    -h, --help                       Show help
+```
+
+Let's continue building our _fabulous_ application adding a new feature:
+by default (i.e. no options given) the application will display the names of the Fab Four. But, if we pass the option `-t` / `--twist` it will display the names in uppercase:
+
+```crystal
+# file: all_my_cli.cr
+require "option_parser"
+
+the_beatles = [
+  "John Lennon",
+  "Paul McCartney",
+  "George Harrison",
+  "Ringo Starr"
+]
+shout = false
+
+option_parser = OptionParser.parse do |parser|
+  parser.banner = "Welcome to The Beatles App!"
+
+  parser.on "-v", "--version", "Please, -please- show me the version" do
+    puts "version 1.0"
+    exit
+  end
+  parser.on "-h", "--help", "Show help" do
+    puts parser
+    exit
+  end
+  parser.on "-t", "--twist", "Twist and SHOUT" do
+    shout = true
+  end
+end
+
+members = the_beatles
+members = the_beatles.map &.upcase if shout
+
+puts ""
+puts "Group members:"
+puts "=============="
+members.each do |member|
+  puts member
+end
+```
+
+Running the application with the `-t` option will output:
+
+```shell-session
+$ crystal run ./all_my_cli.cr -- -t
+
+Group members:
+==============
+JOHN LENNON
+PAUL MCCARTNEY
+GEORGE HARRISON
+RINGO STARR
+```
+
+#### Parameterized options
+Let’s add a new feature to our application: _when passing the option -g, the application will say hello to a given name **passed as a parameter to the option**_.
+
+```crystal
+option_parser = OptionParser.parse do |parser|
+  #...
+
+  parser.on "-g NAME", "--goodbye_hello=NAME", "Say hello to whoever you want" do |name|
+    say_hi_to = name
+  end
+
+  #... etc
+```
+
+In this case, the block receives a parameter that represents the parameter passed to the option.
+
+**Note:** Before running the application, for this new option to be functional, some other changes need to be made. See [the final result](#the-complete-application).
```

Comment:
Maybe the above code example could be a minimal but working application demonstrating the `-g` option.
```cr
say_hi_to = nil
option_parser = OptionParser.parse do |parser|
	  parser.on "-g NAME", "--goodbye_hello=NAME", "Say hello to whoever you want" do |name|
    say_hi_to = name
  end
end
unless say_hi_to.empty?
  puts ""
  puts "You say goodbye, and I say hello to #{say_hi_to}!"
end
```